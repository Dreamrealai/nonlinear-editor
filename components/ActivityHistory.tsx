'use client';

import React, { useState, useEffect, useCallback } from 'react';
import { useSupabase } from '@/components/providers/SupabaseProvider';
import toast from 'react-hot-toast';
import { browserLogger } from '@/lib/browserLogger';
import {
  PAGINATION_CONSTANTS,
  ACTIVITY_TIME_RANGES,
  SPINNER_CONSTANTS,
  FILE_SIZE_CONSTANTS,
} from '@/lib/constants/ui';

interface ActivityMetadata {
  duration?: number;
  resolution?: string;
  aspectRatio?: string;
  fileSize?: number;
  mimeType?: string;
}

interface ActivityHistoryEntry {
  id: string;
  activity_type: string;
  title: string | null;
  description: string | null;
  model: string | null;
  metadata: ActivityMetadata;
  created_at: string;
}

const activityTypeIcons: Record<string, string> = {
  video_generation: 'üé¨',
  audio_generation: 'üéµ',
  image_upload: 'üñºÔ∏è',
  video_upload: 'üìπ',
  audio_upload: 'üéß',
  frame_edit: '‚ú®',
  video_upscale: 'üìà',
};

const activityTypeLabels: Record<string, string> = {
  video_generation: 'Video Generated',
  audio_generation: 'Audio Generated',
  image_upload: 'Image Uploaded',
  video_upload: 'Video Uploaded',
  audio_upload: 'Audio Uploaded',
  frame_edit: 'Frame Edited',
  video_upscale: 'Video Upscaled',
};

/**
 * Memoized activity history entry component
 */
interface ActivityEntryProps {
  entry: ActivityHistoryEntry;
  formatDate: (dateString: string) => string;
}

const ActivityEntry = React.memo<ActivityEntryProps>(function ActivityEntry({
  entry,
  formatDate,
}): React.ReactElement {
  return (
    <div
      key={entry.id}
      className="flex items-start gap-3 rounded-lg border border-neutral-100 p-3 hover:bg-neutral-50 transition-colors"
    >
      <div className="text-2xl flex-shrink-0">{activityTypeIcons[entry.activity_type] || 'üìÑ'}</div>
      <div className="flex-1 min-w-0">
        <div className="flex items-start justify-between gap-2">
          <div className="flex-1 min-w-0">
            <h3 className="text-sm font-medium text-neutral-900 truncate">
              {entry.title || activityTypeLabels[entry.activity_type] || 'Activity'}
            </h3>
            {entry.description && (
              <p className="mt-0.5 text-xs text-neutral-600 line-clamp-2">{entry.description}</p>
            )}
            {entry.model && <p className="mt-0.5 text-xs text-neutral-500">Model: {entry.model}</p>}
            {entry.metadata && Object.keys(entry.metadata).length > 0 && (
              <div className="mt-1 flex flex-wrap gap-2">
                {entry.metadata.duration && (
                  <span className="inline-flex items-center rounded-full bg-neutral-100 px-2 py-0.5 text-xs text-neutral-700">
                    {entry.metadata.duration}s
                  </span>
                )}
                {entry.metadata.resolution && (
                  <span className="inline-flex items-center rounded-full bg-neutral-100 px-2 py-0.5 text-xs text-neutral-700">
                    {entry.metadata.resolution}
                  </span>
                )}
                {entry.metadata.aspectRatio && (
                  <span className="inline-flex items-center rounded-full bg-neutral-100 px-2 py-0.5 text-xs text-neutral-700">
                    {entry.metadata.aspectRatio}
                  </span>
                )}
                {entry.metadata.fileSize && (
                  <span className="inline-flex items-center rounded-full bg-neutral-100 px-2 py-0.5 text-xs text-neutral-700">
                    {(entry.metadata.fileSize / FILE_SIZE_CONSTANTS.BYTES_PER_MB).toFixed(1)} MB
                  </span>
                )}
              </div>
            )}
          </div>
          <span className="text-xs text-neutral-500 whitespace-nowrap">
            {formatDate(entry.created_at)}
          </span>
        </div>
      </div>
    </div>
  );
});

export const ActivityHistory = React.memo(function ActivityHistory(): React.ReactElement {
  const { supabaseClient } = useSupabase();
  const [history, setHistory] = useState<ActivityHistoryEntry[]>([]);
  const [loading, setLoading] = useState(true);
  const [clearLoading, setClearLoading] = useState(false);

  const loadHistory = useCallback(async (): Promise<void> => {
    if (!supabaseClient) return;

    try {
      setLoading(true);
      const response = await fetch(`/api/history?limit=${PAGINATION_CONSTANTS.HISTORY_LIMIT}`);

      if (!response.ok) {
        // Don't show error toast for 500 errors (likely table doesn't exist)
        // Just log it and show empty history
        browserLogger.error(
          { status: response.status, statusText: response.statusText },
          'Error loading activity history'
        );
        setHistory([]);
        setLoading(false);
        return;
      }

      const data = await response.json();
      setHistory(data.history || []);
    } catch (error) {
      browserLogger.error({ error }, 'Error loading activity history');
      // Fail silently - just show empty history
      setHistory([]);
    } finally {
      setLoading(false);
    }
  }, [supabaseClient]);

  const handleClearHistory = useCallback(async (): Promise<void> => {
    const confirmed = confirm(
      'Are you sure you want to clear your entire activity history? This action cannot be undone.'
    );

    if (!confirmed) return;

    setClearLoading(true);

    try {
      const response = await fetch('/api/history', {
        method: 'DELETE',
      });

      if (!response.ok) {
        throw new Error('Failed to clear history');
      }

      setHistory([]);
      toast.success('Activity history cleared');
    } catch (error) {
      browserLogger.error({ error }, 'Error clearing activity history');
      toast.error('Failed to clear activity history');
    } finally {
      setClearLoading(false);
    }
  }, []);

  useEffect((): void => {
    void loadHistory();
  }, [loadHistory]);

  const formatDate = useCallback((dateString: string): string => {
    const date = new Date(dateString);
    const now = new Date();
    const diffMs = now.getTime() - date.getTime();
    const diffMins = Math.floor(diffMs / ACTIVITY_TIME_RANGES.ONE_MINUTE_MS);
    const diffHours = Math.floor(diffMs / ACTIVITY_TIME_RANGES.ONE_HOUR_MS);
    const diffDays = Math.floor(diffMs / ACTIVITY_TIME_RANGES.ONE_DAY_MS);

    if (diffMins < 1) return 'Just now';
    if (diffMins < 60) return `${diffMins} minute${diffMins !== 1 ? 's' : ''} ago`;
    if (diffHours < 24) return `${diffHours} hour${diffHours !== 1 ? 's' : ''} ago`;
    if (diffDays < 7) return `${diffDays} day${diffDays !== 1 ? 's' : ''} ago`;

    return date.toLocaleDateString('en-US', {
      month: 'short',
      day: 'numeric',
      year: date.getFullYear() !== now.getFullYear() ? 'numeric' : undefined,
    });
  }, []);

  return (
    <div className="rounded-xl border border-neutral-200 bg-white p-6 shadow-sm">
      <div className="mb-4 flex items-center justify-between">
        <h2 className="text-xl font-semibold text-neutral-900">Activity History</h2>
        {history.length > 0 && (
          <button
            onClick={handleClearHistory}
            disabled={clearLoading}
            className="rounded-lg bg-neutral-100 px-3 py-1.5 text-xs font-medium text-neutral-700 hover:bg-neutral-200 disabled:cursor-not-allowed disabled:opacity-50"
          >
            {clearLoading ? 'Clearing...' : 'Clear History'}
          </button>
        )}
      </div>

      {loading ? (
        <div className="py-8 text-center">
          <div
            className={`inline-block h-${SPINNER_CONSTANTS.SPINNER_SIZE_MD} w-${SPINNER_CONSTANTS.SPINNER_SIZE_MD} animate-spin rounded-full border-${SPINNER_CONSTANTS.SPINNER_BORDER_WIDTH} border-purple-200 border-t-purple-600 motion-reduce:animate-none dark:border-purple-800 dark:border-t-purple-400`}
            role="status"
            aria-label="Loading history"
          ></div>
          <p className="mt-2 text-sm text-neutral-600 dark:text-neutral-400">Loading history...</p>
        </div>
      ) : history.length === 0 ? (
        <div className="py-12 text-center">
          <div className="mb-2 text-4xl">üìã</div>
          <p className="text-sm text-neutral-600">No activity yet</p>
          <p className="mt-1 text-xs text-neutral-500">
            Your AI generations and uploads will appear here
          </p>
        </div>
      ) : (
        <div className="space-y-3">
          <div className="max-h-96 overflow-y-auto">
            {history.map(
              (entry): React.ReactElement => (
                <ActivityEntry key={entry.id} entry={entry} formatDate={formatDate} />
              )
            )}
          </div>
          <div className="pt-2 text-center text-xs text-neutral-500">
            Showing {history.length} {history.length === 1 ? 'entry' : 'entries'}
          </div>
        </div>
      )}
    </div>
  );
});
