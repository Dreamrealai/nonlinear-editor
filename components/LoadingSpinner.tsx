import React from 'react';
/**
 * LoadingSpinner Component
 *
 * A reusable loading spinner component with customizable size and color.
 * - Uses CSS animations for smooth spinning effect
 * - Optional text label
 * - Multiple size variants
 * - Branded purple gradient variant
 * - Accessibility support with reduced motion
 */

import clsx from 'clsx';

export interface LoadingSpinnerProps {
  /** Size of the spinner */
  size?: 'sm' | 'md' | 'lg' | 'xl';
  /** Custom className for additional styling */
  className?: string;
  /** Text to display next to the spinner */
  text?: string;
  /** Variant for branded or default styling */
  variant?: 'default' | 'branded';
}

const sizeClasses = {
  sm: 'w-4 h-4 border-2',
  md: 'w-6 h-6 border-2',
  lg: 'w-8 h-8 border-3',
  xl: 'w-12 h-12 border-4',
};

export function LoadingSpinner({ size = 'md', className, text, variant = 'default' }: LoadingSpinnerProps): React.ReactElement {
  const borderClasses = variant === 'branded'
    ? 'border-purple-200 border-t-purple-600 dark:border-purple-800 dark:border-t-purple-400'
    : 'border-gray-300 border-t-blue-600 dark:border-gray-700 dark:border-t-blue-400';

  return (
    <div className={clsx('flex items-center gap-2', className)}>
      <div
        className={clsx(
          'animate-spin rounded-full motion-reduce:animate-none motion-reduce:border-t-4',
          borderClasses,
          sizeClasses[size]
        )}
        role="status"
        aria-label="Loading"
      />
      {text && <span className="text-sm text-gray-600 dark:text-gray-400">{text}</span>}
    </div>
  );
}
