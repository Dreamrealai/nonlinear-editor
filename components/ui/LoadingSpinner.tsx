/**
 * LoadingSpinner Component
 *
 * An animated loading indicator with size and variant customization.
 * Uses Lucide's Loader2 icon with continuous spin animation.
 *
 * @example
 * ```tsx
 * <LoadingSpinner />
 * <LoadingSpinner size={32} variant="branded" />
 * <LoadingSpinner size={16} className="text-blue-500" />
 * ```
 */
import { Loader2 } from 'lucide-react';
import { cn } from '@/lib/utils';

/**
 * Props for the LoadingSpinner component
 */
interface LoadingSpinnerProps {
  /** Size of the spinner in pixels (default: 24) */
  size?: number;
  /** Additional CSS classes to apply */
  className?: string;
  /** Visual variant (default or branded with gradient) */
  variant?: 'default' | 'branded';
}

/**
 * A rotating loading spinner icon.
 *
 * Features:
 * - Continuous rotation animation
 * - Customizable size
 * - Branded variant with gradient coloring
 * - Proper ARIA labels for accessibility
 *
 * @param size - Spinner size in pixels (default: 24)
 * @param variant - Visual style variant (default or branded)
 * @param className - Additional CSS classes
 * @returns An animated loading spinner with proper accessibility attributes
 */
export function LoadingSpinner({ size = 24, className, variant = 'default' }: LoadingSpinnerProps) {
  const variantClass = variant === 'branded'
    ? 'bg-gradient-to-r from-purple-500 to-pink-500 bg-clip-text text-transparent'
    : '';

  return (
    <Loader2
      style={{ width: size, height: size }}
      className={cn('animate-spin', variantClass, className)}
      aria-label="Loading"
      role="status"
    />
  );
}
