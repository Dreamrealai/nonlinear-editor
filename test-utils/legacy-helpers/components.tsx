/**
 * Component Test Helpers
 *
 * @deprecated This legacy component helper module is deprecated.
 * Please migrate to modern utilities:
 * - `renderWithProviders()` → Use `render()` from `/test-utils/render.tsx`
 * - `waitForLoadingToFinish()` → Use `waitFor()` from `@testing-library/react`
 * - `setupUserEvent()` → Use `userEvent` from `@testing-library/user-event`
 * - `createMockRouter()` → Available in `/test-utils/testHelpers.ts`
 *
 * **Migration:**
 * ```typescript
 * // OLD:
 * import { renderWithProviders } from '@/test-utils/legacy-helpers/components';
 * const { getByText } = renderWithProviders(<MyComponent />);
 *
 * // NEW:
 * import { render, screen } from '@/test-utils';
 * render(<MyComponent />);
 * const element = screen.getByText('Hello');
 * ```
 *
 * **Migration Guide:** See `/docs/TESTING_UTILITIES.md` section "Component Testing"
 * **Issue:** #83
 *
 * Utilities for testing React components with Testing Library.
 * Provides helpers for rendering with providers, waiting for states,
 * setting up user interactions, and common component patterns.
 *
 * @module __tests__/helpers/components
 * @example
 * ```typescript
 * // DEPRECATED:
 * import { renderWithProviders } from '@/test-utils/legacy-helpers/components';
 *
 * // NEW:
 * import { render, screen } from '@/test-utils';
 * ```
 */

import React, { ReactElement } from 'react';
import { render, RenderOptions, RenderResult, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import type { UserEvent } from '@testing-library/user-event';

/**
 * Options for rendering components with providers.
 */
export interface RenderWithProvidersOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * Initial route for router
   */
  route?: string;

  /**
   * Mock router configuration
   */
  router?: Partial<any>;

  /**
   * Additional wrapper components
   */
  wrapper?: React.ComponentType<{ children: React.ReactNode }>;
}

/**
 * Renders a component with common providers (Router, etc.).
 *
 * This is a wrapper around Testing Library's render that automatically
 * sets up providers commonly needed in tests.
 *
 * @param component - The component to render
 * @param options - Optional render configuration
 * @returns Render result with utilities
 *
 * @example
 * ```typescript
 * const { getByText, user } = renderWithProviders(
 *   <MyComponent />,
 *   { route: '/editor/123' }
 * );
 *
 * await user.click(getByText('Submit'));
 * ```
 */
export function renderWithProviders(
  component: ReactElement,
  options?: RenderWithProvidersOptions
): RenderResult & { user: UserEvent } {
  const { route = '/', router = {}, wrapper, ...renderOptions } = options || {};

  // Setup user event
  const user = userEvent.setup();

  // Create wrapper if provided
  const Wrapper = wrapper
    ? ({ children }: { children: React.ReactNode }) => {
        const CustomWrapper = wrapper;
        return <CustomWrapper>{children}</CustomWrapper>;
      }
    : undefined;

  const result = render(component, {
    wrapper: Wrapper,
    ...renderOptions,
  });

  return {
    ...result,
    user,
  };
}

/**
 * Waits for loading states to finish.
 *
 * Looks for common loading indicators and waits for them to disappear.
 *
 * @param options - Optional waitFor options
 *
 * @example
 * ```typescript
 * renderWithProviders(<MyComponent />);
 * await waitForLoadingToFinish();
 * expect(screen.getByText('Content')).toBeInTheDocument();
 * ```
 */
export async function waitForLoadingToFinish(options?: { timeout?: number }): Promise<void> {
  const { timeout = 5000 } = options || {};

  await waitFor(
    () => {
      // Check for common loading indicators
      const loadingTexts = ['Loading...', 'Please wait...', 'Processing...'];
      const hasLoading = loadingTexts.some((text) => {
        try {
          return document.body.textContent?.includes(text);
        } catch {
          return false;
        }
      });

      expect(hasLoading).toBe(false);
    },
    { timeout }
  );
}

/**
 * Creates a mock Next.js router.
 *
 * @param overrides - Optional router properties to override
 * @returns Mock router object
 *
 * @example
 * ```typescript
 * const mockRouter = createMockRouter({
 *   pathname: '/editor/123',
 *   query: { id: '123' }
 * });
 *
 * jest.mock('next/navigation', () => ({
 *   useRouter: () => mockRouter
 * }));
 * ```
 */
export function createMockRouter(overrides?: Partial<any>): any {
  return {
    push: jest.fn(),
    replace: jest.fn(),
    refresh: jest.fn(),
    back: jest.fn(),
    forward: jest.fn(),
    prefetch: jest.fn().mockResolvedValue(undefined),
    pathname: '/',
    query: {},
    asPath: '/',
    route: '/',
    ...overrides,
  };
}

/**
 * Sets up user event with default options.
 *
 * @param options - Optional user event options
 * @returns UserEvent instance
 *
 * @example
 * ```typescript
 * const user = setupUserEvent();
 * await user.click(button);
 * await user.type(input, 'Hello');
 * ```
 */
export function setupUserEvent(options?: Parameters<typeof userEvent.setup>[0]): UserEvent {
  return userEvent.setup(options);
}

/**
 * Waits for an element to appear and returns it.
 *
 * @param getElement - Function to get the element
 * @param options - Optional wait options
 * @returns The element
 *
 * @example
 * ```typescript
 * const button = await waitForElement(
 *   () => screen.getByRole('button', { name: 'Submit' })
 * );
 * ```
 */
export async function waitForElement<T>(
  getElement: () => T,
  options?: { timeout?: number }
): Promise<T> {
  const { timeout = 3000 } = options || {};
  let element: T;

  await waitFor(
    () => {
      element = getElement();
      expect(element).toBeTruthy();
    },
    { timeout }
  );

  return element!;
}

/**
 * Waits for an element to disappear.
 *
 * @param getElement - Function that should throw when element is gone
 * @param options - Optional wait options
 *
 * @example
 * ```typescript
 * await waitForElementToDisappear(
 *   () => screen.getByText('Loading...')
 * );
 * ```
 */
export async function waitForElementToDisappear(
  getElement: () => any,
  options?: { timeout?: number }
): Promise<void> {
  const { timeout = 3000 } = options || {};

  await waitFor(
    () => {
      expect(() => getElement()).toThrow();
    },
    { timeout }
  );
}

/**
 * Fills a form with values.
 *
 * @param user - UserEvent instance
 * @param fields - Object mapping field names to values
 *
 * @example
 * ```typescript
 * const user = setupUserEvent();
 * await fillForm(user, {
 *   email: 'test@example.com',
 *   password: 'password123'
 * });
 * ```
 */
export async function fillForm(user: UserEvent, fields: Record<string, string>): Promise<void> {
  for (const [name, value] of Object.entries(fields)) {
    const input = document.querySelector(`[name="${name}"]`) as HTMLElement;
    if (!input) {
      throw new Error(`Could not find input with name "${name}"`);
    }
    await user.clear(input);
    await user.type(input, value);
  }
}

/**
 * Clicks a button by its text.
 *
 * @param user - UserEvent instance
 * @param buttonText - Button text or regex
 * @returns Promise that resolves when click completes
 *
 * @example
 * ```typescript
 * const user = setupUserEvent();
 * await clickButton(user, 'Submit');
 * ```
 */
export async function clickButton(user: UserEvent, buttonText: string | RegExp): Promise<void> {
  const button = document.querySelector(`button:contains("${buttonText}")`) as HTMLElement;

  if (!button) {
    // Fallback to finding by role and name
    const buttons = Array.from(document.querySelectorAll('button'));
    const foundButton = buttons.find((btn) => {
      const text = btn.textContent || '';
      return typeof buttonText === 'string' ? text.includes(buttonText) : buttonText.test(text);
    });

    if (!foundButton) {
      throw new Error(`Could not find button with text "${buttonText}"`);
    }

    await user.click(foundButton);
  } else {
    await user.click(button);
  }
}

/**
 * Submits a form.
 *
 * @param user - UserEvent instance
 * @param form - Form element or selector
 *
 * @example
 * ```typescript
 * const user = setupUserEvent();
 * await submitForm(user, 'form[name="login"]');
 * ```
 */
export async function submitForm(user: UserEvent, form: HTMLFormElement | string): Promise<void> {
  const formElement =
    typeof form === 'string' ? (document.querySelector(form) as HTMLFormElement) : form;

  if (!formElement) {
    throw new Error(`Could not find form: ${form}`);
  }

  const submitButton = formElement.querySelector('[type="submit"]') as HTMLElement;
  if (submitButton) {
    await user.click(submitButton);
  } else {
    // Simulate form submission
    formElement.dispatchEvent(new Event('submit', { bubbles: true, cancelable: true }));
  }
}

/**
 * Waits for a specific number of milliseconds.
 *
 * @param ms - Milliseconds to wait
 *
 * @example
 * ```typescript
 * await wait(1000); // Wait 1 second
 * ```
 */
export function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Expects an element to have specific CSS classes.
 *
 * @param element - The element to check
 * @param classes - Expected class names
 *
 * @example
 * ```typescript
 * const button = screen.getByRole('button');
 * expectToHaveClasses(button, 'bg-blue-500', 'text-white');
 * ```
 */
export function expectToHaveClasses(element: HTMLElement, ...classes: string[]): void {
  classes.forEach((className) => {
    expect(element).toHaveClass(className);
  });
}

/**
 * Expects an element to be visible and enabled.
 *
 * @param element - The element to check
 *
 * @example
 * ```typescript
 * const button = screen.getByRole('button');
 * expectToBeInteractive(button);
 * ```
 */
export function expectToBeInteractive(element: HTMLElement): void {
  expect(element).toBeVisible();
  expect(element).toBeEnabled();
}

/**
 * Expects an element to be disabled.
 *
 * @param element - The element to check
 *
 * @example
 * ```typescript
 * const button = screen.getByRole('button');
 * expectToBeDisabled(button);
 * ```
 */
export function expectToBeDisabled(element: HTMLElement): void {
  expect(element).toBeDisabled();
}

/**
 * Expects text content to match.
 *
 * @param element - The element to check
 * @param text - Expected text or pattern
 *
 * @example
 * ```typescript
 * expectTextContent(heading, 'Welcome');
 * expectTextContent(paragraph, /hello/i);
 * ```
 */
export function expectTextContent(element: HTMLElement, text: string | RegExp): void {
  if (typeof text === 'string') {
    expect(element).toHaveTextContent(text);
  } else {
    expect(element.textContent).toMatch(text);
  }
}

/**
 * Gets all error messages from the page.
 *
 * @returns Array of error message elements
 *
 * @example
 * ```typescript
 * const errors = getErrorMessages();
 * expect(errors).toHaveLength(1);
 * expect(errors[0]).toHaveTextContent('Invalid input');
 * ```
 */
export function getErrorMessages(): HTMLElement[] {
  return Array.from(document.querySelectorAll('[role="alert"], .error, [class*="error"]'));
}

/**
 * Expects no error messages on the page.
 *
 * @example
 * ```typescript
 * await submitForm(user, form);
 * expectNoErrors();
 * ```
 */
export function expectNoErrors(): void {
  const errors = getErrorMessages();
  expect(errors).toHaveLength(0);
}

/**
 * Expects specific error message to be present.
 *
 * @param message - Expected error message or pattern
 *
 * @example
 * ```typescript
 * expectErrorMessage('Password is required');
 * expectErrorMessage(/required/i);
 * ```
 */
export function expectErrorMessage(message: string | RegExp): void {
  const errors = getErrorMessages();
  const found = errors.some((error) => {
    const text = error.textContent || '';
    return typeof message === 'string' ? text.includes(message) : message.test(text);
  });

  expect(found).toBe(true);
}

/**
 * Mocks window.alert.
 *
 * @returns Mock function
 *
 * @example
 * ```typescript
 * const alertMock = mockAlert();
 * // ... trigger alert
 * expect(alertMock).toHaveBeenCalledWith('Success!');
 * ```
 */
export function mockAlert(): jest.Mock {
  const mock = jest.fn();
  global.alert = mock;
  return mock;
}

/**
 * Mocks window.confirm.
 *
 * @param returnValue - What confirm should return (default: true)
 * @returns Mock function
 *
 * @example
 * ```typescript
 * const confirmMock = mockConfirm(true);
 * // ... trigger confirm
 * expect(confirmMock).toHaveBeenCalled();
 * ```
 */
export function mockConfirm(returnValue: boolean = true): jest.Mock {
  const mock = jest.fn().mockReturnValue(returnValue);
  global.confirm = mock;
  return mock;
}

/**
 * Mocks window.prompt.
 *
 * @param returnValue - What prompt should return
 * @returns Mock function
 *
 * @example
 * ```typescript
 * const promptMock = mockPrompt('User input');
 * // ... trigger prompt
 * expect(promptMock).toHaveBeenCalled();
 * ```
 */
export function mockPrompt(returnValue: string | null = null): jest.Mock {
  const mock = jest.fn().mockReturnValue(returnValue);
  global.prompt = mock;
  return mock;
}
