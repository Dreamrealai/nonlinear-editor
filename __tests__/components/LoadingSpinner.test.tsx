import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { LoadingSpinner } from '@/components/LoadingSpinner';

describe('LoadingSpinner', () => {
  describe('Rendering', () => {
    it('should render without crashing', () => {
      const { container } = render(<LoadingSpinner />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toBeInTheDocument();
    });

    it('should have loading aria-label', () => {
      const { container } = render(<LoadingSpinner />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveAttribute('aria-label', 'Loading');
    });

    it('should have role status for accessibility', () => {
      const { container } = render(<LoadingSpinner />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toBeInTheDocument();
    });
  });

  describe('Size Variants', () => {
    it('should render small size', () => {
      const { container } = render(<LoadingSpinner size="sm" />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveClass('w-4', 'h-4', 'border-2');
    });

    it('should render medium size by default', () => {
      const { container } = render(<LoadingSpinner />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveClass('w-6', 'h-6', 'border-2');
    });

    it('should render large size', () => {
      const { container } = render(<LoadingSpinner size="lg" />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveClass('w-8', 'h-8', 'border-3');
    });

    it('should render extra large size', () => {
      const { container } = render(<LoadingSpinner size="xl" />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveClass('w-12', 'h-12', 'border-4');
    });
  });

  describe('Text Display', () => {
    it('should display text when provided', () => {
      render(<LoadingSpinner text="Loading..." />);
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('should not display text when not provided', () => {
      const { container } = render(<LoadingSpinner />);
      const textElement = container.querySelector('span');
      expect(textElement).not.toBeInTheDocument();
    });

    it('should style text correctly', () => {
      render(<LoadingSpinner text="Please wait" />);
      const textElement = screen.getByText('Please wait');
      expect(textElement).toHaveClass('text-sm', 'text-gray-600');
    });

    it('should display custom text', () => {
      render(<LoadingSpinner text="Custom loading message" />);
      expect(screen.getByText('Custom loading message')).toBeInTheDocument();
    });
  });

  describe('Custom Styling', () => {
    it('should accept custom className', () => {
      const { container } = render(<LoadingSpinner className="custom-class" />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('custom-class');
    });

    it('should combine custom className with default classes', () => {
      const { container } = render(<LoadingSpinner className="mt-4" />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('flex', 'items-center', 'gap-2', 'mt-4');
    });

    it('should apply multiple custom classes', () => {
      const { container } = render(<LoadingSpinner className="mt-4 mb-4 mx-auto" />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('mt-4', 'mb-4', 'mx-auto');
    });
  });

  describe('Animation', () => {
    it('should have spin animation class', () => {
      const { container } = render(<LoadingSpinner />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveClass('animate-spin');
    });

    it('should have rounded-full class', () => {
      const { container } = render(<LoadingSpinner />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveClass('rounded-full');
    });

    it('should have correct border colors', () => {
      const { container } = render(<LoadingSpinner />);
      const spinner = container.querySelector('[role="status"]');
      expect(spinner).toHaveClass('border-gray-300', 'border-t-blue-600');
    });
  });

  describe('Layout', () => {
    it('should have flex layout', () => {
      const { container } = render(<LoadingSpinner />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('flex');
    });

    it('should align items center', () => {
      const { container } = render(<LoadingSpinner />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('items-center');
    });

    it('should have gap between spinner and text', () => {
      const { container } = render(<LoadingSpinner text="Loading" />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('gap-2');
    });
  });

  describe('Combined Props', () => {
    it('should render with all props', () => {
      const { container } = render(
        <LoadingSpinner size="lg" text="Processing..." className="my-custom-class" />
      );

      const wrapper = container.firstChild;
      const spinner = container.querySelector('[role="status"]');

      expect(wrapper).toHaveClass('my-custom-class');
      expect(spinner).toHaveClass('w-8', 'h-8');
      expect(screen.getByText('Processing...')).toBeInTheDocument();
    });

    it('should work with size and text', () => {
      const { container } = render(<LoadingSpinner size="sm" text="Loading..." />);
      const spinner = container.querySelector('[role="status"]');

      expect(spinner).toHaveClass('w-4', 'h-4');
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('should work with size and className', () => {
      const { container } = render(<LoadingSpinner size="xl" className="mt-8" />);
      const spinner = container.querySelector('[role="status"]');
      const wrapper = container.firstChild;

      expect(spinner).toHaveClass('w-12', 'h-12');
      expect(wrapper).toHaveClass('mt-8');
    });
  });

  describe('Edge Cases', () => {
    it('should handle empty text prop', () => {
      const { container } = render(<LoadingSpinner text="" />);
      const textElement = container.querySelector('span');
      expect(textElement).not.toBeInTheDocument();
    });

    it('should handle whitespace text', () => {
      render(<LoadingSpinner text="   " />);
      const whitespaceSpans = screen.getAllByText((_, element) => element?.textContent === '   ');
      expect(whitespaceSpans.length).toBeGreaterThan(0);
    });

    it('should handle undefined className gracefully', () => {
      const { container } = render(<LoadingSpinner className={undefined} />);
      const wrapper = container.firstChild;
      expect(wrapper).toHaveClass('flex', 'items-center', 'gap-2');
    });

    it('should handle long text', () => {
      const longText = 'This is a very long loading message that might wrap to multiple lines';
      render(<LoadingSpinner text={longText} />);
      expect(screen.getByText(longText)).toBeInTheDocument();
    });
  });

  describe('Snapshot Consistency', () => {
    it('should maintain structure with default props', () => {
      const { container } = render(<LoadingSpinner />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should maintain structure with all props', () => {
      const { container } = render(
        <LoadingSpinner size="lg" text="Loading data..." className="custom-spinner" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
