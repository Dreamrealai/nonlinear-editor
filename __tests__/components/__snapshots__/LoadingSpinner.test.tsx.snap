// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoadingSpinner Snapshot Consistency should maintain structure with all props 1`] = `
<div
  class="flex items-center gap-2 custom-spinner"
>
  <div
    aria-label="Loading"
    class="animate-spin rounded-full motion-reduce:animate-none motion-reduce:border-t-4 border-gray-300 border-t-blue-600 dark:border-gray-700 dark:border-t-blue-400 w-8 h-8 border-3"
    role="status"
  />
  <span
    class="text-sm text-gray-600 dark:text-gray-400"
  >
    Loading data...
  </span>
</div>
`;

exports[`LoadingSpinner Snapshot Consistency should maintain structure with default props 1`] = `
<div
  class="flex items-center gap-2"
>
  <div
    aria-label="Loading"
    class="animate-spin rounded-full motion-reduce:animate-none motion-reduce:border-t-4 border-gray-300 border-t-blue-600 dark:border-gray-700 dark:border-t-blue-400 w-6 h-6 border-2"
    role="status"
  />
</div>
`;
